File: dawn_haskell/src/Fusion_tree.hs
data FusionTree a = Leaf a | Node (FusionTree a) (FusionTree a)

empty :: Num a => FusionTree a
empty = Leaf 0

isEmpty :: FusionTree a -> Bool
isEmpty (Leaf 0) = True
isEmpty _ = False

insert :: Num a => a -> FusionTree a -> FusionTree a
insert x (Leaf 0) = Leaf x
insert x (Leaf y) = Node (Leaf x) (Leaf y)
insert x (Node left right) = Node (insert x left) right

find :: Num a => a -> FusionTree a -> a
find x (Leaf y) = y
find x (Node left right) = find x left + find x right

union :: FusionTree a -> FusionTree a -> FusionTree a
union (Leaf 0) t = t
union t (Leaf 0) = t
union (Leaf x) (Leaf y) = Node (Leaf x) (Leaf y)
union (Node ll lr) (Node rl rr) = Node (union ll rl) (union lr rr)

File: dawn_go/fusion_tree.go
type FusionTree struct {
    value int
    left  *FusionTree
    right *FusionTree
}

func NewEmpty() *FusionTree {
    return &FusionTree{value: 0}
}

func (t *FusionTree) IsEmpty() bool {
    return t.value == 0 && t.left == nil && t.right == nil
}

func (t *FusionTree) Insert(x int) *FusionTree {
    if t.IsEmpty() {
        return &FusionTree{value: x}
    }
    left := t.left.Insert(x)
    right := t.right
    return &FusionTree{value: t.value, left: left, right: right}
}

func (t *FusionTree) Find(x int) int {
    if t.IsEmpty() {
        return 0
    }
    if t.value == x {
        return t.value + t.left.Find(x) + t.right.Find(x)
    }
    return t.left.Find(x) + t.right.Find(x)
}

func Union(left, right *FusionTree) *FusionTree {
    if left.IsEmpty() {
        return right
    }
    if right.IsEmpty() {
        return left
    }
    return &FusionTree{
        value: left.value + right.value,
        left:  Union(left.left, right.left),
        right: Union(left.right, right.right),
    }
}